### PR TITLE
fix the TF stuff

### PR DESCRIPTION
**Pull request by yosoufe**
Resolving this problem: https://edstem.org/us/courses/4944/discussion/411771?comment=974035 , 

and some typo

### CODE REVIEW

**Comment by yosoufe** [comment]
If you need more discussions, we can set up a call maybe.

File: packages/my_static_tf/src/static_at_tf_publisher_map1.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def fetch_tag_id(tag) -> int:
 MAP_TRANSFORM_DICT = {
     31: ((0, 0, 0.075), (0, 0, 0)),
     32: ((0, 0.5, 0.075), (0, 0, 0)),
-    65: ((0.5, 0.25, 0), (0, 0, 0)),
+    65: ((0.5, 0.25, 0.075), (0, 0, 0)),
```
Comment (fishbotics) [inline packages/my_static_tf/src/static_at_tf_publisher_map1.py:22]: @yosoufe, can you explain why you changed this value?

File: packages/my_at_localization/src/my_at_pose_node.py [hunk 2]
```diff
@@ -149,15 +159,15 @@ def _broadcast_detected_tag(self, image_msg, tag_id, tag):
         self._broadcast_tf(
             parent_frame_id=tag_frame_id,
             child_frame_id=tag_cam_frame_id,
-            euler_angles=(-90 * math.pi / 180, 0, 90 * math.pi / 180)
+            euler_angles=(-90 * math.pi / 180, 0, -90 * math.pi / 180)
```
Comment (fishbotics) [inline packages/my_at_localization/src/my_at_pose_node.py:162]: These are now the same value twice. Is that what you intended?

File: packages/my_static_tf/src/static_at_tf_publisher_map1.py (hunk 1 above)
Comment (rohjunha) [reply packages/my_static_tf/src/static_at_tf_publisher_map1.py:22]: I think this is fixing the mistake. Each tuple defines the transformation of AprilTags; the first component is the translation and the second component is rotation. I think I omitted the z-axis offset of AprilTag #65.

File: packages/my_at_localization/src/my_at_pose_node.py [hunk 3]
```diff
@@ -137,6 +137,16 @@ def _broadcast_tf(
     def _broadcast_detected_tag(self, image_msg, tag_id, tag):
         pose = np.identity(4)
         pose[0:3, 0:3] = tag.pose_R
+        # inverse the rotation and tranformation comming out of the AT detector.
+        # The AP detector's output is the camera frame relative to the TAG
+        # According to: https://duckietown.slack.com/archives/C6ZHPV212/p1619876209086300?thread_ts=1619751267.084600&cid=C6ZHPV212
+        # While the transformation below needs TAG relative to camera
+        # Therefore we need to reverse it first.
+        pose = np.transpose(pose) # reverse orientation
+        pose_t = np.ones((4,1))
```
Comment (fishbotics) [inline packages/my_at_localization/src/my_at_pose_node.py:146]: `np.ones((4, 1))`

File: packages/my_at_localization/src/my_at_pose_node.py [hunk 4]
```diff
@@ -137,6 +137,16 @@ def _broadcast_tf(
...
         pose = np.identity(4)
         pose[0:3, 0:3] = tag.pose_R
+        # inverse the rotation and tranformation comming out of the AT detector.
+        # The AP detector's output is the camera frame relative to the TAG
+        # According to: https://duckietown.slack.com/archives/C6ZHPV212/p1619876209086300?thread_ts=1619751267.084600&cid=C6ZHPV212
+        # While the transformation below needs TAG relative to camera
+        # Therefore we need to reverse it first.
+        pose = np.transpose(pose) # reverse orientation
+        pose_t = np.ones((4,1)) 
+        pose_t[:3,:] = tag.pose_t
```
Comment (fishbotics) [inline packages/my_at_localization/src/my_at_pose_node.py:147]: `pose_t[:3, :] = tag.pose_t`

File: packages/my_at_localization/src/my_at_pose_node.py [hunk 5]
```diff
@@ -137,6 +137,16 @@ def _broadcast_tf(
...
+        pose_t = - np.matmul(pose,pose_t) # reverse translation
```
Comment (fishbotics) [inline packages/my_at_localization/src/my_at_pose_node.py:148]: `pose_t = - np.matmul(pose, pose_t)`

File: packages/my_at_localization/src/my_at_pose_node.py [hunk 6]
```diff
@@ -137,6 +137,16 @@ def _broadcast_tf(
...
+        pose_t = pose_t[:3,:]
```
Comment (fishbotics) [inline packages/my_at_localization/src/my_at_pose_node.py:149]: `pose_t = pose_t[:3, :]`

File: packages/my_static_tf/src/static_at_tf_publisher_map1.py (hunk 1 above)
Comment (yosoufe) [reply packages/my_static_tf/src/static_at_tf_publisher_map1.py:22]: This is  not related to fixing the TFs. 

We already discussed this in one of the office hours. Why one of the tags location has `z=0` and the others have `z=0.075` ? I think all of them should be at the height of `0.075`. Correct?

Comment (fishbotics) [reply packages/my_static_tf/src/static_at_tf_publisher_map1.py:22]: Oh yeah you're right. When it's all in the same PR, I get confused.

File: packages/my_at_localization/src/my_at_pose_node.py (hunk 2 above)
Comment (yosoufe) [reply packages/my_at_localization/src/my_at_pose_node.py:162]: If you mean twice of `-90 * math.pi / 180`, then yes.

File: packages/my_static_tf/src/static_at_tf_publisher_map1.py (hunk 1 above)
Comment (yosoufe) [reply packages/my_static_tf/src/static_at_tf_publisher_map1.py:22]: sorry, I had to clarify it.

**Comment by fishbotics** [comment]
@rohjunha you should approve since I am not as familiar with this file.

File: packages/my_at_localization/src/my_at_pose_node.py (hunk 6 above)
Comment (yosoufe) [inline packages/my_at_localization/src/my_at_pose_node.py:149]: How I inversed the transformation matrix: https://math.stackexchange.com/a/152686

Probably `np.linalg.inv` would work.. Not tested.

**Comment by rohjunha** [comment]
> @rohjunha you should approve since I am not as familiar with this file.

Yes, I am still looking at this. I have run this code on my robot and seems to work as @yosoufe showed in the video.

**Comment by rohjunha** [comment]
It took some time to check whether applying `np.linalg.inv` is different from the original update.
I found some errors when I applied `np.linalg.inv`, but then I realized that this happens to the original update as well.

> detected 31: (1620067450.2974062, 1620067457.7160935)                                                                                                                                             [27/2335]
> detected 65: (1620067450.2974062, 1620067457.7209518)
> Error, more than one new minimum found.                                                              
> detected 32: (1620067450.3350942, 1620067457.7755446)
> detected 31: (1620067450.3350942, 1620067457.780984)
> detected 65: (1620067450.3350942, 1620067457.7863784)
> Error, more than one new minimum found.                                                              
> detected 32: (1620067450.3697257, 1620067457.8370004)
> detected 31: (1620067450.3697257, 1620067457.843041)
> detected 65: (1620067450.3697257, 1620067457.8482084)

Probably, this is not coming from the matrix inversion, but coming from the apriltag.
https://github.com/AprilRobotics/apriltag/issues/59

**Pull request merged**